packages/kbn-grid-layout/grid/grid_panel/grid_panel.tsx

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import React, { forwardRef, useEffect, useMemo } from 'react';
+import { combineLatest, skip } from 'rxjs';
+
+import { EuiPanel, euiFullHeight, useEuiOverflowScroll } from '@elastic/eui';
+import { css } from '@emotion/react';
+import { euiThemeVars } from '@kbn/ui-theme';
+import { GridLayoutStateManager, PanelInteractionEvent } from '../types';
+import { getKeysInOrder } from '../utils/resolve_grid_row';
+import { DragHandle } from './drag_handle';
+import { ResizeHandle } from './resize_handle';
+
+export interface GridPanelProps {
+  panelId: string;
+  rowIndex: number;
+  renderPanelContents: (panelId: string) => React.ReactNode;
+  interactionStart: (
+    type: PanelInteractionEvent['type'] | 'drop',
+    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
+  ) => void;
+  gridLayoutStateManager: GridLayoutStateManager;
+}
+
+export const GridPanel = React.memo(
```

Comment:
1. I think `useMemo` on `renderPanelContents` is more efficient here - otherwise, we call `renderPanelContents` whenever one of the props changes, which isn't necessary. We should only refetch the panel contents when the panel ID changes - this is especially important when the panels are embeddables.

    Here's what the interaction looks like on drag (without changing sections, just dragging and dropping within the same section) on my branch:

     
     ![image](https://github.com/user-attachments/assets/3e5fa0d6-a453-442f-a744-5a6cefe8b0ef)

    
    Versus your branch:
    
     ![image](https://github.com/user-attachments/assets/96799950-29c1-4df5-8fa1-896595330c51)

2 & 3: Good call on this :+1:

Reply:
I'm actually wondering if the additional renders might be coming from the HTML structure changing..... I will need to do some testing here.

Reply:
huh, that makes sense. I think I tested it out before adding the onDrop sync callback but once it's added, it doesn't make sense. I reverted the changes, thanks for the explanation!